### ISSUE TITLE
Slim v4.0 parser failure

### ISSUE THREAD

**Issue by Jakenberg** [presidentbeef/brakeman#1284]
### Background

Brakeman version: 4.3.1
Rails version: 5.0.2
Ruby version: 2.3.7

Link to Rails application code: N/A

#### Parse Error

The following error is produced by running `brakeman -d` in project directory:
```
== Errors ==

Error: Expected tag   [project_folder]/app/views/layouts/application.html.slim, Line 16, Column 62     javascript(nonce==content_security_policy_script_nonce):                                                             ^
Location: [project_folder]/app/views/layouts/application.html.slim:16
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/slim-3.0.7/lib/slim/parser.rb:513:in `syntax_error!'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/slim-3.0.7/lib/slim/parser.rb:367:in `parse_tag'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/slim-3.0.7/lib/slim/parser.rb:248:in `parse_line_indicators'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/slim-3.0.7/lib/slim/parser.rb:192:in `parse_line'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/slim-3.0.7/lib/slim/parser.rb:95:in `call'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/temple-0.7.7/lib/temple/engine.rb:50:in `block in call'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/temple-0.7.7/lib/temple/engine.rb:50:in `each'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/temple-0.7.7/lib/temple/engine.rb:50:in `inject'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/temple-0.7.7/lib/temple/engine.rb:50:in `call'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/temple-0.7.7/lib/temple/mixins/template.rb:10:in `compile'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/temple-0.7.7/lib/temple/templates/tilt.rb:30:in `prepare'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bundle/ruby/2.5.0/gems/tilt-2.0.8/lib/tilt/template.rb:99:in `initialize'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/parsers/template_parser.rb:94:in `new'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/parsers/template_parser.rb:94:in `parse_slim'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/parsers/template_parser.rb:29:in `parse_template'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/scanner.rb:87:in `block in parse_files'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/file_parser.rb:27:in `block in read_files'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/file_parser.rb:26:in `each'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/file_parser.rb:26:in `read_files'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/scanner.rb:86:in `parse_files'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/scanner.rb:45:in `process'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman.rb:354:in `scan'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman.rb:77:in `run'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/commandline.rb:133:in `run_brakeman'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/commandline.rb:118:in `regular_report'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/commandline.rb:142:in `run_report'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/commandline.rb:35:in `run'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/lib/brakeman/commandline.rb:20:in `start'
[home_folder]/.rvm/gems/ruby-2.3.7/gems/brakeman-4.3.1/bin/brakeman:8:in `<top (required)>'
[home_folder]/.rvm/gems/ruby-2.3.7/bin/brakeman:23:in `load'
[home_folder]/.rvm/gems/ruby-2.3.7/bin/brakeman:23:in `<main>'
[home_folder]/.rvm/gems/ruby-2.3.7/bin/ruby_executable_hooks:24:in `eval'
[home_folder]/.rvm/gems/ruby-2.3.7/bin/ruby_executable_hooks:24:in `<main>'
```

Support for "tags" within `javascript`/`css` selectors was added in v4 of `slim`. It appears that the latest version of brakeman is using `slim` v3.

Some context... `content_security_policy_script_nonce` is a view helper from the [secure headers gem](https://github.com/twitter/secure_headers).

**Comment by Jakenberg** [comment]
I created a [pull request](https://github.com/presidentbeef/brakeman/pull/1285) where I upgraded the slim version and ran through the brakeman test suite. Everything appeared rainbow (nice touch). 

This is blocking my team so anything to lower the burden on you guys to merge would greatly benefit us. As such, I'm happy to help with any additional testing needs if desired.

**Comment by presidentbeef** [comment]
Thanks @Jakenberg!

Honestly, the automated tests do not have a lot of Slim in them, and I'm not sure about backwards compatibility between 3->4. Have you tested the changes on your own codebase?

**Comment by Jakenberg** [comment]
@presidentbeef I'm very green on brakeman's codebase so its hard for me to say. This is what I found in slim's changelog:
```
4.0.0 (2018-08-26)

  * Add ability to specify attributes for embedded engines using TagEngine - #653
  * Set Tilt's default encoding to utf-8 - #800
  * Prevent generation of invalid Ruby code in heredoc - #770
  * Don't convert underscores to dashes on data attributes - #807 [Backward incompatible]
  * Update description in README.jp.md - #795
  * Removed support for embedded engines: Asciidoc, Builder, Creole, Erb, Nokogiri, Opal, Org-Ruby, Styl, Wikicloth
  * Fix tests
```

_Possibly_ the default encoding change to Tilt; conversion of underscores to dashes; removal of a certain embedded engine's support could produce issues.

In our case, we are no longer seeing the brakeman error 👍 My gut is telling me that, generally speaking, any issues with backwards compatibility could at least be resolved by the fact that whomever is using the brakeman gem still has control over what specific version of slim is used during the parsing step via their Gemfile. It is just that now the availability of _what_ version is elevated to the current one.

**Comment by presidentbeef** [comment]
I agree, looks like 4.0 doesn't break anything with respect to Brakeman.

> any issues with backwards compatibility could at least be resolved by the fact that whomever is using the brakeman gem still has control over what specific version of slim is used during the parsing step via their Gemfile.

This is not 100% accurate, as the main `brakeman` gem bundles a snapshot of all dependencies. But if people want more control over specific versions, the `brakeman-lib` and `brakeman-min` gems do allow for that.

**Comment by presidentbeef** [comment]
Closed with #1285
